### PR TITLE
Fix Restoration from Seed (Old)

### PR DESCRIPTION
**Pull request by RyanRory**

### CODE REVIEW

File: SignalServiceKit/src/Messages/OWSMessageDecrypter.m [hunk 1]
```diff
@@ -682,6 +684,8 @@ - (void)notifyUserForErrorMessage:(TSErrorMessage *)errorMessage
 {
     NSString *hexEncodedPublicKey = [LKDatabaseUtilities getMasterHexEncodedPublicKeyFor:envelope.source in:transaction] ?: envelope.source;
     TSThread *contactThread = [TSContactThread getOrCreateThreadWithContactId:hexEncodedPublicKey transaction:transaction];
+    contactThread.shouldThreadBeVisible = true;
```
Comment (nielsandriesse) [inline SignalServiceKit/src/Messages/OWSMessageDecrypter.m:687]: Why does the thread need to be visible?

File: SignalServiceKit/src/Messages/OWSMessageDecrypter.m [hunk 2]
```diff
@@ -669,8 +669,10 @@ - (void)processException:(NSException *)exception envelope:(SSKProtoEnvelope *)e
 
         OWSAssertDebug(errorMessage);
         if (errorMessage != nil) {
-            [errorMessage saveWithTransaction:transaction];
             [LKSessionManagementProtocol handleDecryptionError:errorMessage.errorType forHexEncodedPublicKey:envelope.source using:transaction];
+            if (![LKSessionManagementProtocol isErrorMessageBeforeRestoration:errorMessage]) {
```
Comment (nielsandriesse) [inline SignalServiceKit/src/Messages/OWSMessageDecrypter.m:673]: I wouldn't categorize this as session management. Also we already have `isFriendRequestFromBeforeRestoration(...)` which basically does the same thing. It'd be better to just modify that and move it to `SessionMetaProtocol`.

File: SignalServiceKit/src/Messages/OWSMessageDecrypter.m [hunk 3]
```diff
@@ -669,8 +669,10 @@ - (void)processException:(NSException *)exception envelope:(SSKProtoEnvelope *)e
...
+                [errorMessage saveWithTransaction:transaction];
+            }
             [self notifyUserForErrorMessage:errorMessage envelope:envelope transaction:transaction];
```
Comment (nielsandriesse) [inline SignalServiceKit/src/Messages/OWSMessageDecrypter.m:676]: Do we still want to do this even if the error was from before restoration?

File: SignalServiceKit/src/Loki/Protocol/Session Management/SessionManagementProtocol.swift [hunk 4]
```diff
@@ -234,6 +242,8 @@ public final class SessionManagementProtocol : NSObject {
         let publicKey = thread.contactIdentifier()
         print("[Loki] End session message received from: \(publicKey).")
         // Notify the user
+        let masterHexEncodedPublicKey = storage.getMasterHexEncodedPublicKey(for: publicKey, in: transaction) ?? publicKey
+        let thread = TSContactThread.getOrCreateThread(withContactId: masterHexEncodedPublicKey, transaction: transaction)
```
Comment (nielsandriesse) [inline SignalServiceKit/src/Loki/Protocol/Session Management/SessionManagementProtocol.swift:246]: Why is this needed?

File: SignalServiceKit/src/Loki/Protocol/Session Management/LokiSessionResetImplementation.swift [hunk 5]
```diff
@@ -45,19 +46,28 @@ public class LokiSessionResetImplementation : NSObject, SessionResetProtocol {
...
         // If the current user initiated the reset then send back an empty message to acknowledge the completion of the session reset
-        if thread.sessionResetStatus == .initiated {
+        if let masterThread = TSContactThread.getWithContactId(masterHexEncodedPublicKey, transaction: transaction), masterThread.sessionResetStatus == .initiated {
             let emptyMessage = EphemeralMessage(thread: thread)
-            SSKEnvironment.shared.messageSender.sendPromise(message: emptyMessage).retainUntilComplete()
+            SSKEnvironment.shared.messageSender.sendPromise(message: emptyMessage).done2{ _ in
+                // Clear the session reset status
+                masterThread.sessionResetStatus = .none
+                masterThread.save()
+            }.retainUntilComplete()
```
Comment (nielsandriesse) [inline SignalServiceKit/src/Loki/Protocol/Session Management/LokiSessionResetImplementation.swift:61]: `retainUntilComplete()` shouldn't be necessary in Swift

File: SignalServiceKit/src/Loki/Protocol/Session Management/LokiSessionResetImplementation.swift [hunk 6]
```diff
@@ -45,19 +46,28 @@ public class LokiSessionResetImplementation : NSObject, SessionResetProtocol {
...
         guard let thread = TSContactThread.getWithContactId(recipientID, transaction: transaction) else {
             Logger.debug("[Loki] A new session was adopted but the thread couldn't be found for: \(recipientID).")
             return
         }
         // If the current user initiated the reset then send back an empty message to acknowledge the completion of the session reset
-        if thread.sessionResetStatus == .initiated {
+        if let masterThread = TSContactThread.getWithContactId(masterHexEncodedPublicKey, transaction: transaction), masterThread.sessionResetStatus == .initiated {
             let emptyMessage = EphemeralMessage(thread: thread)
-            SSKEnvironment.shared.messageSender.sendPromise(message: emptyMessage).retainUntilComplete()
+            SSKEnvironment.shared.messageSender.sendPromise(message: emptyMessage).done2{ _ in
```
Comment (nielsandriesse) [inline SignalServiceKit/src/Loki/Protocol/Session Management/LokiSessionResetImplementation.swift:57]: This'll work but it'd be better to use the message sender job queue to ensure that things still proceed even if the user closes the app before the message is sent

File: SignalServiceKit/src/Loki/Protocol/Multi Device/MultiDeviceProtocol.swift [hunk 7]
```diff
@@ -30,6 +30,25 @@ public final class MultiDeviceProtocol : NSObject {
 
     // MARK: - Initialization
     private override init() { }
+    
+    // MARK: - Recover Linked devices
```
Comment (nielsandriesse) [inline SignalServiceKit/src/Loki/Protocol/Multi Device/MultiDeviceProtocol.swift:34]: We can just put this under general

File: SignalServiceKit/src/Loki/Protocol/Meta/SessionMetaProtocol.swift [hunk 8]
```diff
@@ -58,9 +58,16 @@ public final class SessionMetaProtocol : NSObject {
     @objc(isThreadNoteToSelf:)
     public static func isThreadNoteToSelf(_ thread: TSThread) -> Bool {
         guard let thread = thread as? TSContactThread else { return false }
+        let hexEncodedPublicKey = thread.contactIdentifier()
         var isNoteToSelf = false
         storage.dbReadConnection.read { transaction in
-            isNoteToSelf = LokiDatabaseUtilities.isUserLinkedDevice(thread.contactIdentifier(), transaction: transaction)
+            //When session reset status is not none, only send the message to linked device
+            let masterHexEncodedPublicKey = storage.getMasterHexEncodedPublicKey(for: hexEncodedPublicKey, in: transaction) ?? hexEncodedPublicKey
+            if let masterThread = TSContactThread.getWithContactId(masterHexEncodedPublicKey, transaction: transaction), masterThread.sessionResetStatus == .none {
```
Comment (nielsandriesse) [inline SignalServiceKit/src/Loki/Protocol/Meta/SessionMetaProtocol.swift:66]: Mixing in session reset logic here is bad. This method should just tell you whether the given thread is a note to self. If at the call site the session reset logic matters you should probably just handle that case manually.

**Comment by nielsandriesse** [comment]
You'll probably want to pull in #222 as I resolved the conflicts there

File: SignalServiceKit/src/Messages/OWSMessageDecrypter.m (hunk 1 above)
Comment (RyanRory) [reply SignalServiceKit/src/Messages/OWSMessageDecrypter.m:687]: When recovering from seed, if we get rid of all the error messages from messages before restoration, the thread won't show up in HomeVC.

File: SignalServiceKit/src/Loki/Protocol/Session Management/SessionManagementProtocol.swift (hunk 4 above)
Comment (RyanRory) [reply SignalServiceKit/src/Loki/Protocol/Session Management/SessionManagementProtocol.swift:246]: To make sure the session reset in progress message is shown in the master thread. Or if this is triggered by a secondary device, the info message seems to be missing.

File: SignalServiceKit/src/Loki/Protocol/Meta/SessionMetaProtocol.swift (hunk 8 above)
Comment (RyanRory) [reply SignalServiceKit/src/Loki/Protocol/Meta/SessionMetaProtocol.swift:66]: Yeah, I was thinking of that when doing this on Android. It is confusing to put the logic here.

File: SignalServiceKit/src/Messages/OWSMessageDecrypter.m (hunk 1 above)
Comment (nielsandriesse) [reply SignalServiceKit/src/Messages/OWSMessageDecrypter.m:687]: Okay, but what if an error occurs in a thread that should be invisible?

File: SignalServiceKit/src/Loki/Protocol/Session Management/SessionManagementProtocol.swift (hunk 4 above)
Comment (nielsandriesse) [reply SignalServiceKit/src/Loki/Protocol/Session Management/SessionManagementProtocol.swift:246]: That makes sense :)

**Comment by nielsandriesse** [comment]
This PR has been superseded
